### PR TITLE
More robust sync mechanism

### PR DESCRIPTION
**Pull request by mbyczkowski**
This PR should address issues discussed in #30.

**TL;DR: This PR introduces retries with exponential backoff (+ jitter for spreading the load) for each client sync API call**

Thoughts/questions:
- I was on the fence where the retry should happen (in syncer or client), but went with the client to keep the complexity hidden.
- Currently, we ignore fail counts as long as a retry succeeded. We might care at some point to see how often we have to retry to succeed, but seemed like unnecessary noise at this point.
- ~~I'm still pondering the best config representation for the backoff, and I'm leaning towards using strings and `time.ParseDuration`. I've added some simple validation checks for the config, so would be easy to get this changed and handle potential errors.~~ `Timeout`, `MinBackoff` and `MaxBackoff` can now be specified in the same way as `PollInterval` (so `20s`, `100ms` etc.)
- I've relied on a simple 3rd party lib for backoff (that I have used before). It's MIT licensed, so ~~hopefully~~ that's OK.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -92,9 +96,17 @@ func (c KeywhizHTTPClient) Logger() *logrus.Entry {
 
 // NewClient produces a read-to-use client struct given PEM-encoded certificate file, key file, and
```
Comment (stfinney) [inline client.go]: nit: this comment should be updated to reflect the change in arguments

File: client.go [hunk 2]
```diff
@@ -273,3 +298,29 @@ func (p httpClientParams) buildClient() (*http.Client, error) {
...
+		//These are the defaults
+		Min:    c.params.minBackoff,
+		Max:    c.params.maxBackoff,
+		Jitter: true,
+	}
+
+	for i := 0; i < c.params.maxRetries; i++ {
+		now := time.Now()
+		resp, err = c.httpClient.Get(t.String())
+		if err != nil || resp.StatusCode == http.StatusOK {
```
Comment (mcpherrinm) [inline client.go]: I'm wondering if there's any other status codes we shouldn't retry for.  404 maybe.

File: config.go [hunk 3]
```diff
@@ -76,6 +85,22 @@ func LoadConfig(configFile string) (*Config, error) {
...
+	if config.ClientTimeout == "" {
+		config.ClientTimeout = "60s"
+	}
+
+	if config.MinBackoff == "" {
+		config.MinBackoff = "100ms"
+	}
+
+	if config.MaxBackoff == "" {
+		config.MaxBackoff = "10000ms"
```
Comment (mcpherrinm) [inline config.go]: nit: "10s" is easier to read

File: config.go [hunk 4]
```diff
@@ -45,11 +50,15 @@ type Config struct {
...
-	Group   string `yaml:"group"`     // Optional: If unspecified, the global defaults are used.
+	Key        string `yaml:"key"`         // Mandatory: Path to PEM key to use
+	Cert       string `yaml:"cert"`        // Optional: PEM Certificate (If cert isn't in key file)
+	User       string `yaml:"user"`        // Optional: User and Group are defaults for files without metadata
+	DirName    string `yaml:"directory"`   // Optional: What directory under SecretsDir this client is in. Defaults to the client name.
+	Group      string `yaml:"group"`       // Optional: If unspecified, the global defaults are used.
+	MaxRetries uint16 `yaml:"max_retries"` // Optional: retry each HTTP call after non-200 response. Defaults to Keysync's config.
+	Timeout    string `yaml:"timeout"`     // Optional: timeout HTTP calls. Defaults to Keysync's config.
+	MinBackoff string `yaml:"min_backoff"` // Optional: wait time before first retry. Defaults to Keysync's config.
+	MaxBackoff string `yaml:"max_backoff"` // Optional: max wait time before retries. Defaults to Keysync's config.
```
Comment (mcpherrinm) [inline config.go]: I don't think we need to be able to configure backoff per-client, and it's just extra complexity.  We don't allow configuring servers per-client, and it makes sense to configure retries and backoff with the server.

File: client.go (hunk 2 above)
Comment (mbyczkowski) [reply client.go]: probably 401, 403 and generally most (if not all 4xx)? I think we should mainly care about 5xx. I will pull this into `isIntermittent` method
